embedded/store/kv_metadata.go

```diff
@@ -36,22 +40,60 @@ func (md *KVMetadata) Deleted() bool {
 	return md.deleted
 }
 
+func (md *KVMetadata) ExpiresAt(expiresAt *time.Time) *KVMetadata {
+	md.expiresAt = expiresAt
+	return md
+}
+
+func (md *KVMetadata) Expirable() bool {
+	return md.expiresAt != nil
+}
+
+func (md *KVMetadata) ExpirationTime() *time.Time {
+	return md.expiresAt
+}
+
+func (md *KVMetadata) ExpiredAt(mtime time.Time) bool {
+	if !md.Expirable() {
+		return false
+	}
+
+	return !md.expiresAt.After(mtime)
+}
+
 func (md *KVMetadata) Bytes() []byte {
-	var b byte
+	var b [1 + tsSize]byte
 
 	if md.deleted {
-		b = b | deletedFlag
+		b[0] = 1
+	}
+
+	if md.expiresAt == nil {
+		return b[:1]
 	}
 
-	return []byte{b}
+	binary.BigEndian.PutUint64(b[1:], uint64(md.expiresAt.Unix()))
+
+	return b[:]
 }
 
 func (md *KVMetadata) ReadFrom(b []byte) error {
 	if len(b) == 0 {
 		return nil
 	}
 
-	md.deleted = b[0]&deletedFlag != 0
+	md.deleted = b[0] != 0
```

Comment:
I think the first byte should still be a bit field - one for deleted, other for expiration date.
Otherwise it will be hard to extend the metadata with new fields.

Reply:
good point, implemented a generic solution to add new attributes in the future